### PR TITLE
Asynchronous gyro and acc processing

### PR DESCRIPTION
**Pull request by DzikuVx**
This PR replaces #409 and #369 and it will introduce unified way of handling both features with one configuration option
- [x] CLI interface
- [x] tasks and task integration
- [x] filter initialization with proper frequencies
- [x] tests
- [x] docs

### CODE REVIEW

File: src/main/blackbox/blackbox.c [hunk 1]
```diff
@@ -1284,7 +1281,7 @@ static bool blackboxWriteSysinfo()
             }
             );
 
-        BLACKBOX_PRINT_HEADER_LINE("looptime:%d",                           targetLooptime);
+        BLACKBOX_PRINT_HEADER_LINE("looptime:%d",                           getLooptime());
```
Comment (digitalentity) [inline src/main/blackbox/blackbox.c]: Looptime now refers to PID looptime or gyro update interval?

Comment (DzikuVx) [inline src/main/blackbox/blackbox.c]: It is still PID task. This would have to be extended too, as well as configurator

**Comment by martinbudden** [comment]
@DzikuVx , have you seen @borisbstyle 's work on running the gyro update off of an interrupt? See https://github.com/betaflight/betaflight/tree/gyro_on_exti

**Comment by DzikuVx** [comment]
@martinbudden Yes, I know he is working on smth like that but never actually dived into into. Maybe it's time to catch up

**Comment by DzikuVx** [comment]
I've decided to rewrite this branch :)

**Comment by digitalentity** [comment]
@DzikuVx looks good so far!

**Comment by DzikuVx** [comment]
Looks flyable on a bench. Wanted to test in flight this weekend but instead decided to give betaflight 3 a chance to compare how far behind we are in terms of acro :)

**Comment by digitalentity** [comment]
Great! As for comparison with Betaflight - I'm also very curious. It's subjective, but my QAV/INAV flies equally well as BF in acro and much nicer and smoother in ANGLE/HORIZON.

**Comment by DzikuVx** [comment]
So, 10 batteries with Betaflight 3 and I think BF flies slightly better. Seems somewhat smoother, but difference is not that big as few months ago, so I would say, we did a pretty good job here catching up

**Comment by digitalentity** [comment]
Great! I think notch filter and 2-dof PID will get us to 90% of betaflight performance :smile:

**Comment by DzikuVx** [comment]
Flight tested with #609 using 2kHz/1kHz/90Hz/60Hz (gyro/PID/acc/atti) and it flies just great!
No horizon drift observed.
@digitalentity you wanted to introduce some kind of precomputing for ATTI

**Comment by digitalentity** [comment]
Yes, in the next few days I'll try to find some time to modify your code for better ATTI calculation, expect a PR against your branch :smile:

**Comment by DzikuVx** [comment]
I'm done with development here, you can also commit here if you want

**Comment by digitalentity** [comment]
Thanks! I'm kind of busy these days, so please don't expect instantaneous changes.

**Comment by DzikuVx** [comment]
rebased against development
@martinbudden please make me a pleasure and do not change too much in the core files before this is merged :)

**Comment by martinbudden** [comment]
OK. I'm working on the RX code at the moment, so that shouldn't effect you work. But I'll avoid any changes to core files.

**Comment by digitalentity** [comment]
The last commit makes `taskAcc` and `taskGyro` slightly heavier, but also makes `taskAttitude` more accurate.

Also, position estimation is heavily affected by attitude calculation rate, so going below 50-100Hz on attitude is probably going to affect position estimation considerably. This is something worth looking into and I'll probably have the time to think about it next week.

**Comment by digitalentity** [comment]
Currently position estimation is coupled with PID loop, I think it's worth splitting it to a separate task as well or combining it with ATTI as it's more logical.

This PR is going to be a big change, we'll need to make sure async modes work for navigation as well.

**Comment by martinbudden** [comment]
@DzikuVx , can I ask you to make a small change:

For "attitude" you sometimes use "atti" and sometimes "attitude". Using the abbreviation doesn't really bring any benefits and also makes the code slightly less readable and maintainable. Can you use "attitude" throughout?

[One of "Martin's rules of programming": "no unnecessary abbreviations".]

**Comment by DzikuVx** [comment]
@martinbudden I suppose I can do it :)

**Comment by DzikuVx** [comment]
@digitalentity 

> The last commit makes taskAcc and taskGyro slightly heavier, but also makes taskAttitude more accurate.

Which commit?

**Comment by digitalentity** [comment]
My commit, latest one.

**Comment by digitalentity** [comment]
Oops, looks like I didn't push. Sorry

**Comment by DzikuVx** [comment]
Note: SONAR support for F1 removed, after merging update hardware map

**Comment by DzikuVx** [comment]
I really, really, really do not want to rebase it again :)
Either way, looks like it works fine after rebase on a workbench

**Comment by DzikuVx** [comment]
Will test rabase in a few hours

**Comment by DzikuVx** [comment]
Bench tested with all configurations after latest rebase, @digitalentity let's merge it finally :)

**Comment by stronnag** [comment]
I would prefer it had been flight tested and shown to work well with navigation modes, which it did not last time I tried it at least.

**Comment by stronnag** [comment]
I can try the aysnc gyro branch again this weekend. Are there any recommendations for settings (in particular task priorities) for nav modes ?

**Comment by DzikuVx** [comment]
By default, async processing is disabled.

For nav modes I recommend setting `async_mode = GYRO` and combination of `gyro_lpf` and `gyro_sync_denom` as low as target allows. 188HZ and gyro_sync_denom=1 should be enough
In this case accelometer computations would stay binded to PID loop, only gyro would run faster.

**Comment by stronnag** [comment]
Thanks. I'll give it  a try on an F3 this weekend.

**Comment by digitalentity** [comment]
@stronnag thanks!

`async_mode = NONE` shouldn't change anything.

I recommend testing `async_mode = GYRO` with looptime=2000/3000/4000 and 

`async_mode = ALL` with different `acc_task_frequency` (suggested values 200,500,1000) and `attitude_task_frequency` (suggested values 100,200,500)

**Comment by digitalentity** [comment]
I think I'll merge this right away. We can improve it later on. @all, any objections?

**Comment by martinbudden** [comment]
No objections from me. As you say, once it is in we can improve it.

**Comment by digitalentity** [comment]
@DzikuVx merged! Now comes the fun part - make it all configurable from GUI :wink:

**Comment by DzikuVx** [comment]
Will not be that bad.... probably only a new INAV specific MSP frame

**Comment by stronnag** [comment]
I flew gyro-sync today, in horizon and navigation modes.

The test craft is well tested and stable. The first flight in the log is on 1.3.0 and is the test basis for subsequent tests:
- Take off, assess horizon mode drift with no stick inputs
- PH
- Mission
- RTH
- Optionally, some free flight

The first log entry is 1.3.0. For the first 20+ seconds, with no stick inputs the craft maintains a reasonably stable position.

It then accurately flies a mission, PH and flies acceptably if boringly and executes RTH, disarming efficiently and without 'bounce'.

Ignore log 2 (1.3.0, no useful context)

Logs 3,4,5 are 1.4.0 (async_mode = GYRO); 3 & 4 show uncommanded drift backwards at 1 or 2 m/s with no stick input (I had not realised that 1.4.0 also changed modes, resulting in my WP switching becoming FAILSAFE (which worked as RTH / failsafe).

Log 6 is 1.4.0 (async_mode = GYRO) shows uncommanded forward drift without stick command, accelerating to 4m/s before WP is engaged. The WP mission is not flown as accurately as previously (the little 360 turn at label WP 11 for example); as a result of lack of confidence in the firmware and low battery, RTH was initiated. The craft failed to disarm and after some bouncing around on the ground is manually disarmed.

Log 7 is 1.4.0 (async_mode = NONE) shows uncommanded forward drift without stick command, accelerating to 3m/s before PH is engaged. The mission is then initiated. It is flown better than Log 6, but not as well as Log 1. RTH is imitated and the machine lands and disarms efficiently.

The tests were conducted within a 30 minute period, with comparable environmental factors (wind, temperature etc). I appreciate there was a slight degradation of satellite coverage between logs 1 and 6,7 and that one test is not necessarily definitive.

**Summary**. A regression against 1.3.0. The uncommanded movements are disconcerting and the self-level / navigation / land / disarm performance in (async_mode = GYRO) is unacceptably poor; the performance with (async_mode = NONE) is still a regression, especially the uncommanded drift. In my view, this has been merged prematurely and may prejudice testing of other 1.4.0 features. I'll reluctant to fly it again and would not wish to fly it on a more valuable platform.

**References:**
**Blackbox log:** http://seyrsnys.myzen.co.uk/2016-11-05/bb.TXT
**1.4.0 CLI dump:** http://seyrsnys.myzen.co.uk/2016-11-05/dc_1000_async_fp.txt
**Video replays:**
2016-11-05 log1 => https://vimeo.com/190378128
2016-11-05 log6 => https://vimeo.com/190378127
2016-11-05 log7 => https://vimeo.com/190378126

**Comment by DzikuVx** [comment]
That is interesting... `async_mode = NONE` leaves processing just like in 1.3. I suspect something else got in between. Will take a look soon

**Comment by stronnag** [comment]
I was surprised about that too.

**Comment by DzikuVx** [comment]
@stronnag Weather here is very poor, so flying in my case is rather not possible ATM.
Could you build and test 1.4 with `#define ASYNC_GYRO_PROCESSING` commented out from target/common.h
This completely disables async_gyro

**Comment by stronnag** [comment]
Of course, it's just unseasonable cold here. I will try that tomorrow.

**Comment by DzikuVx** [comment]
I some inconstitency here. Blackbox log says you have gyro_lpf = OFF while dump states 188HZ. What was is after all?

**Comment by DzikuVx** [comment]
@stronnag could also provide 1.3 CLI dump? I want to compare something

**Comment by stronnag** [comment]
The first log (1.3.0) was with gyro_lpf = 256HZ, the 1.4.0 logs are gyro_lpf = 188HZ and have gyro_lpf=1 in the log header. I had a recollection it was @digitalentity's recommendation to use 188 ...

**Comment by stronnag** [comment]
1.3.0 CLI dump: http://seyrsnys.myzen.co.uk/2016-11-05/dc_1000_fp.txt

I can test 1.4.0 at both. I (maybe erroneously) assumed 188 would be more stable.

**Comment by DzikuVx** [comment]
Could you concentrate on `aync_mode=NONE` tomorrow? I want to know what causes degraded performance with _disabled_ async processing.

**Comment by stronnag** [comment]
Sure. Let me know what different settings would be useful to test. I can easily reflash and reconfigure in the field.

**Comment by DzikuVx** [comment]
Right now I'm interested only in async_mode = NONE. So, basically, settings
from 1.3 and defaults from 1.4. remember about flight modes :)

sob., 5.11.2016 o 20:27 użytkownik stronnag notifications@github.com
napisał:

**Comment by DzikuVx** [comment]
@stronnag @digitalentity see #761

**Comment by stronnag** [comment]
I will try imu-use-downsampled-gyro with my 1.3 settings and async=NONE first ..

**Comment by digitalentity** [comment]
Hm... this is odd. There is one piece of code that _might_ affect performance even with async=NONE. I'll try to investigate todya.

**Comment by digitalentity** [comment]
@stronnag, @DzikuVx no need to test anything.

Regression confirmed. IMU for some reason doesn't correct attitude from accelerometer or does so incorrectly. Attitude does drift with time hence the drift and degraded nav performance. 

Happens with any `async_mode` setting.

**Comment by digitalentity** [comment]
Bug found. Fix pending.

**Comment by digitalentity** [comment]
@DzikuVx, @stronnag this bug appears to be my doing - missed one line of code in `imuUpdateMeasuredAcceleration()` function when compiling with `ASYNC_GYRO_PROCESSING`. See #762 for a fix.

**Pull request merged**